evm/src/maverick-v2/MaverickV2SwapAdapter.sol

```diff
@@ -0,0 +1,334 @@
+    /// @param tickLimit The tick limit for the swap.
+    /// @param specifiedAmount The amount of the token to buy.
+    /// @return calculatedAmount The amount of the token bought.
+    function buy(
+        IMaverickV2Pool pool,
+        bool isTokenAIn,
+        int32 tickLimit,
+        uint256 specifiedAmount
+    ) internal returns (uint256 calculatedAmount) {
+        IMaverickV2Pool.SwapParams memory swapParams = IMaverickV2Pool
+            .SwapParams({
+            amount: specifiedAmount,
+            tokenAIn: isTokenAIn,
+            exactOutput: true,
+            tickLimit: tickLimit
+        });
+        // callback data is the sender address
+        bytes memory data = abi.encode(msg.sender);
+        (uint256 amountIn,) = pool.swap(msg.sender, swapParams, data);
+        return amountIn;
+    }
+
+    /// @notice Sell tokens to a pool.
+    /// @param pool The pool to sell to.
+    /// @param sellToken The token to sell.
+    /// @param isTokenAIn Whether token A is the input token.
+    /// @param tickLimit The tick limit for the swap.
+    /// @param specifiedAmount The amount of the token to sell.
+    /// @return calculatedAmount The amount of the token sold.
+    function sell(
+        IMaverickV2Pool pool,
+        address sellToken,
+        bool isTokenAIn,
+        int32 tickLimit,
+        uint256 specifiedAmount
+    ) internal returns (uint256 calculatedAmount) {
+        IMaverickV2Pool.SwapParams memory swapParams = IMaverickV2Pool
+            .SwapParams({
+            amount: specifiedAmount,
+            tokenAIn: isTokenAIn,
+            exactOutput: false,
+            tickLimit: tickLimit
+        });
+        // Pay the pool with the sell token
+        pay(IERC20(sellToken), msg.sender, address(pool), specifiedAmount);
+        (, uint256 amountOut) = pool.swap(msg.sender, swapParams, "");
+        return amountOut;
+    }
+
+    /// @notice MaverickV2SwapCallback is the callback function for MaverickV2
+    /// pools.
+    /// @param tokenIn The token being swapped.
+    /// @param amountIn The amount of the token being swapped.
+    /// @param data The data passed to the callback.
+    function maverickV2SwapCallback(
+        IERC20 tokenIn,
+        uint256 amountIn,
+        uint256,
+        bytes calldata data
+    ) external {
+        require(
+            factory.isFactoryPool(IMaverickV2Pool(msg.sender)), "NotFactoryPool"
+        );
+        address payer = abi.decode(data, (address));
+        pay(tokenIn, payer, msg.sender, amountIn);
+    }
+
+    /// @notice Pay a recipient with a token.
+    /// @param token The token to pay with.
+    /// @param payer The payer of the token.
+    /// @param recipient The recipient of the token.
+    /// @param value The amount of the token to pay.
+    function pay(IERC20 token, address payer, address recipient, uint256 value)
+        internal
+    {
+        if (IWETH9(address(token)) == weth && address(this).balance >= value) {
```

Comment:
Maverick doesn't seem to support native ETH, only WETH (please correct me if I'm wrong) - in this case, I think we shouldn't add Native ETH support on the adapter contract, as this adds unnecessary overhead.

Reply:
Thanks! Removed as unnecessary.